### PR TITLE
Transition effects may now produce zero, one, or multiple messages

### PR DESCRIPTION
**Pull request by fsoikin**
Say good bye to your `Nop` messages with `forkVoid` and `forkMaybe`, and say hello to progress reporting with `forks`!

It has been an occasional complaint for quite some time that "_`init` (or `update`) does not get a `dispatch`, so we can't produce deferred effects from it_".

The `Transition` data structure now models effects as callback-taking rather than message-returning. This allows effects to produce zero, one, or multiple messages. To provide a simpler API for simple cases, our good old friend `fork` (which is still a thing) got three new friends - `forkVoid` for fire-and-forget effects, `forkMaybe` - for effects that may or may not produce a message, and `forks` - for the full power of producing zero, one, or more effects via a callback.

I hope `forkVoid` and `forkMaybe` would be self-expanatory, but for `forks` I have added a new example - see https://github.com/collegevine/purescript-elmish-examples/pull/21, which demonstrates reporting progress of a long-running operation:

inc-slowly

### CODE REVIEW

File: src/Elmish/Component.purs [hunk 1]
```diff
@@ -70,6 +78,15 @@ instance trBind :: Bind (Transition m msg) where
         let (Transition s' cmds') = f s
         in Transition s' (cmds <> cmds')
 
+-- | Smart constructor for the `Transition` type. See comments there. This
+-- | function takes the new (i.e. update) state and an array of commands - i.e.
+-- | effects producing messages, - and constructs a `Transition` out of them
```
Comment (mcordova47) [inline src/Elmish/Component.purs:83]:
```suggestion
-- | effects producing messages - and constructs a `Transition` out of them
```

File: src/Elmish/Component.purs [hunk 2]
```diff
@@ -53,12 +55,18 @@ import Elmish.Trace (traceTime)
...
+data Transition m msg state = Transition state (Array (Command m msg))
+
+-- | An effect that is launched as a result of a component state transition.
+-- | It's a function that takes a callback that allows it to produce (aka
+-- | "dispatch") messages.
+type Command m msg = (msg -> Effect Unit) -> m Unit
 
 instance trBifunctor :: Functor m => Bifunctor (Transition m) where
-    bimap f g (Transition s cmds) = Transition (g s) (f <$$> cmds)
+    bimap f g (Transition s cmds) = Transition (g s) (cmds <#> \cmd sink -> cmd $ sink <<< f)
```
Comment (mcordova47) [inline src/Elmish/Component.purs:69]: Where does the name `sink` come from?

Comment (fsoikin) [reply src/Elmish/Component.purs:69]: In engineering "sink" is a traditional name for a device that takes stuff away somewhere. Compare with "heat sink" or "money sink". In software this term becomes specialized to a mechanism that takes events or messages and sends them somewhere, we don't care where. See https://en.m.wikipedia.org/wiki/Sink_(computing)

Comment (mcordova47) [reply src/Elmish/Component.purs:69]: > Even your kitchen sink kinda does that if you sink about it.

🤦

**Pull request merged**